core/src/main/java/org/apache/calcite/sql/type/SqlTypeUtil.java

```diff
@@ -1015,17 +1015,23 @@ private static boolean flattenFields(
    * @param type         type descriptor
    * @param charSetName  charSet name
    * @param maxPrecision The max allowed precision.
+   * @param allowsPrecisionUnspecified whether allows precision not specified
+   * @param defaultPrecision The default precision.
    * @return corresponding parse representation
    */
   public static SqlDataTypeSpec convertTypeToSpec(RelDataType type,
-      String charSetName, int maxPrecision) {
+      String charSetName, int maxPrecision,
+      boolean allowsPrecisionUnspecified, int defaultPrecision) {
     SqlTypeName typeName = type.getSqlTypeName();
```

Comment:
I would suggest to replace `allowsPrecisionUnspecified` and `defaultPrecision` with a single `explicitPrecision`, and if a `explicitPrecision` is specified, use it directly.

Reply:
This pull request has been updated, thanks for you reviewing, it's really helpful.